File: core/projection.py
"""Turns a player + opponent + line into a projection.

This is the orchestration layer: it pulls the player's game log, builds the
recency-weighted window, asks core.monte_carlo_sim for the matchup adjustment,
and hands both to one of the models in core.projection_models.

The models themselves are closed-form, so a projection is a few hundred
microseconds of arithmetic rather than a 10,000-draw sampling run. That makes
it cheap enough to evaluate *every* model on every request, which is what
powers the model-consensus view - if the parametric fit and the player's own
empirical history disagree about a line, that disagreement is a genuinely
useful signal and worth showing rather than hiding behind one number.
"""

import logging

import numpy as np

from core.data_loader import find_player, find_player_career
from core.ml_model import features_for_next_game, get_model
from core.monte_carlo_sim import create_weight, STAT_MAP
from core.projection_models import (
    MAX_WINDOW, MODELS, DEFAULT_MODEL, recency_weights, run_model, weighted_moments,
)

# Windows the hit-rate breakdown is reported over. "How often has he actually
# cleared this number?" is a different - and more directly checkable - question
# than what any model predicts, so it's worth showing alongside.
HIT_RATE_WINDOWS = [('last_3', 3), ('last_5', 5), ('last_10', 10), ('season', None), ('career', None)]


def hit_rates(name, stat_cat, line):
    """How often the player has reached `line`, over several lookbacks.

    Counted from actual game logs, not modelled - if a receiver has cleared 60
    yards in 4 of his last 10 and 38% of his career, that is a fact about what
    happened, and it belongs next to the projection rather than behind it.
    """
    career = find_player_career(name)
    if career.empty or stat_cat not in career.columns:
        return []

    values = career[stat_cat].fillna(0.0)
    current_season = career['season'].max()
    season_mask = (career['season'] == current_season).to_numpy()

    out = []
    for key, window in HIT_RATE_WINDOWS:
        if key == 'career':
            sample = values.to_numpy(dtype=float)
        elif key == 'season':
            sample = values.to_numpy(dtype=float)[season_mask]
        else:
            sample = values.to_numpy(dtype=float)[-window:]

        games = int(len(sample))
        if games == 0:
            continue
        hits = int(np.sum(sample >= line))
        out.append({
            'window': key,
            'games': games,
            'hits': hits,
            'rate': hits / games,
            'average': float(np.mean(sample)),
        })
    return out


def _window(name, stat_cat):
    """The player's most recent games for `stat_cat`, oldest -> newest."""
    df = find_player(name)
    if stat_cat not in df.columns:
        raise ValueError(f"'{stat_cat}' has no recorded data for {name}")

    series = df.sort_values('week')[stat_cat].dropna()
    values = series.tail(MAX_WINDOW).to_numpy(dtype=float)
    if len(values) == 0:
        raise ValueError(f"Not enough recent games for {name} to run a projection")
    return values


def project(name, def_team, stat_cat, line, model=DEFAULT_MODEL):
    """Project `stat_cat` for `name` against `def_team` and price `line`.

    Returns the chosen model's read plus every other model's over probability,
    so the caller can show how sensitive the answer is to the assumed shape.
    """
    if stat_cat not in STAT_MAP:
        raise ValueError(f"Unsupported stat category '{stat_cat}'")

    values = _window(name, stat_cat)
    weights = recency_weights(len(values))
    shift = create_weight(name, def_team, stat_cat)
    raw_mean, _var, ess = weighted_moments(values, weights)

    alternatives = {
        key: round(run_model(key, values, weights, line, stat_cat, shift).prob_over, 6)
        for key in MODELS
    }

    ml = _ml_projection(name, def_team, stat_cat, line)
    if ml is not None:
        alternatives['ml'] = round(ml['prob_over'], 6)

    if model == 'ml':
        if ml is None:
            raise ValueError(
                f"The trained model isn't available for {stat_cat} - not enough "
                "history for this player or stat."
            )
        projection, prob_over, std = ml['projection'], ml['prob_over'], ml['std_dev']
        model_key, model_label = 'ml', ml['label']
    else:
        result = run_model(model, values, weights, line, stat_cat, shift)
        projection, prob_over, std = result.projection, result.prob_over, result.std
        model_key = result.model if model == DEFAULT_MODEL else model
        model_label = result.label

    return {
        'projection': projection,
        'prob_over': prob_over,
        'prob_under': 1.0 - prob_over,
        'weight': shift,
        'model': model_key,
        'model_label': model_label,
        'form_average': float(raw_mean),
        'season_average': float(np.mean(values)),
        'recent_games': int(len(values)),
        'effective_games': float(ess),
        'std_dev': std,
        'window_games': int(len(values)),
        'alternatives': alternatives,
        'hit_rates': hit_rates(name, stat_cat, line),
        'ml_projection': None if ml is None else ml['projection'],
    }


def _ml_projection(name, def_team, stat_cat, line):
    """The trained model's read, or None when it can't be applied."""
    try:
        trained = get_model(stat_cat)
        if trained is None:
            return None
        features = features_for_next_game(name, def_team, stat_cat, trained)
        if features is None:
            return None
        prediction = float(trained.predict(features)[0])
        return {
            'projection': prediction,
            'prob_over': trained.prob_over(prediction, line),
            'std_dev': trained.spread(prediction),
            'label': 'Trained ridge regression',
        }
    except Exception:  # noqa: BLE001 - never let the ML path break a projection
        logging.getLogger(__name__).exception("Trained model failed for %s / %s", name, stat_cat)
        return None

File: frontend/src/engine/projection.ts
/**
 * The projection pipeline, ported from core/projection.py and
 * core/monte_carlo_sim.create_weight.
 *
 * Given a player's game log and the precomputed league aggregates, this
 * reproduces what the Python engine returned: a recency-weighted window, a
 * matchup adjustment, every model's read on the line, and the hit rates.
 *
 * Parity with Python is enforced by fixtures — see engine.test.ts.
 */

import type { HitRate, HitRateWindow, ProjectionResponse } from '../api/types'
import type { Aggregates, GameRow, ModelsFile, TrainedModelFile } from './bundle'
import { OFFSEASON_GAP_GAMES, decayWeights, runModel, weightedMoments } from './models'
import { clamp, quantileFromSorted } from './numerics'

const MODEL_KEYS = ['ensemble', 'lognormal', 'negbin', 'empirical', 'triangular'] as const

/**
 * A model's P(over) through its calibration map, as core/calibration.py
 * apply_map() does it: p' = sigmoid(a * logit(p) + b). No map, no change.
 */
export function calibrated(models: ModelsFile, stat: string, key: string, p: number): number {
  const ab = models.calibration?.[stat]?.[key]
  if (!ab) return p
  const q = clamp(p, 1e-4, 1 - 1e-4)
  const z = ab[0] * Math.log(q / (1 - q)) + ab[1]
  return 1 / (1 + Math.exp(-z))
}

/** Sample standard deviation (pandas .std() default, ddof=1). */
function sampleStd(values: number[]): number {
  if (values.length < 2) return NaN
  const mean = values.reduce((a, b) => a + b, 0) / values.length
  const ss = values.reduce((acc, v) => acc + (v - mean) ** 2, 0)
  return Math.sqrt(ss / (values.length - 1))
}

function statValues(games: GameRow[], stat: string): number[] {
  return games
    .map((g) => g[stat])
    .filter((v): v is number => typeof v === 'number' && !Number.isNaN(v))
}

/**
 * The player's rolling form — their last `rollingGames` games, across seasons.
 * Mirrors find_player(): last season's games stay in the list until this
 * season's push them out, one game at a time, so nobody starts September with
 * an empty history.
 */
function formGames(games: GameRow[], rollingGames: number): GameRow[] {
  return chronological(games).slice(-rollingGames)
}

function chronological(games: GameRow[]): GameRow[] {
  return [...games].sort((a, b) => Number(a.season) - Number(b.season) || Number(a.week) - Number(b.week))
}

/**
 * The matchup adjustment, ported from create_weight().
 *
 * Quarterbacks are scored against the opponent's team-level pass/run defense
 * relative to the league; skill players against what that defense allows to
 * players of the same depth-chart rank. Both paths read only from precomputed
 * aggregates, which is why they fit in the browser.
 */
export function createWeight(
  games: GameRow[],
  opponent: string,
  stat: string,
  aggregates: Aggregates,
): number {
  const { constants } = aggregates
  const form = formGames(games, constants.rolling_games)
  if (form.length === 0) return 0

  const position = String(form[form.length - 1].position ?? '')
  const mapping = constants.stat_map[stat]
  if (!mapping) return 0

  const playerStd = sampleStd(statValues(form, stat))
  let weight: number

  if (position === 'QB') {
    const k = constants.position_k.QB
    const league = aggregates.league_team_stats[stat]
    const defense = aggregates.defense_weekly[opponent]
    if (!league || !defense || !league.std) return 0

    const rows = mapping[1] === 'pass' ? defense.pass : defense.run
    const allowed = rows.map((r) => r[stat]).filter((v) => typeof v === 'number' && !Number.isNaN(v))
    if (allowed.length === 0) return 0

    const teamAvg = allowed.reduce((a, b) => a + b, 0) / allowed.length
    const zDefense = (teamAvg - league.mean) / league.std
    weight = k * playerStd * -zDefense
  } else {
    // Position-level, scaled by how much of the gap actually repeats. This
    // replaced a depth-chart-rank comparison that measured out as noise; see
    // core/monte_carlo_sim.py for the numbers.
    const k = constants.position_k[position] ?? constants.default_k
    const leagueAvg = aggregates.position_allowed[`NFL|${position}|${stat}`]
    const defenseAvg = aggregates.position_allowed[`${opponent}|${position}|${stat}`]
    const reliability = aggregates.signal_reliability[`${position}|${stat}`]
    if (leagueAvg == null || defenseAvg == null || reliability == null) return 0

    weight = k * reliability * (defenseAvg - leagueAvg)
  }

  return Number.isFinite(weight) ? weight : 0
}

// ---------------------------------------------------------------------------
// Hit rates
// ---------------------------------------------------------------------------

/**
 * How often the player actually reached the line, over several lookbacks.
 *
 * Counted from the game log rather than modelled — the most directly checkable
 * thing on the page.
 */
export function hitRates(games: GameRow[], stat: string, line: number): HitRate[] {
  const values = games.map((g) => {
    const v = g[stat]
    return typeof v === 'number' && !Number.isNaN(v) ? v : 0
  })
  // "Season" means this player's most recent season, which for someone who
  // hasn't played this year is not the league's current one. Matches the
  // Python, which reads the max season off the player's own career.
  const latestSeason = games.reduce((max, g) => Math.max(max, Number(g.season) || 0), 0)
  const seasonMask = games.map((g) => Number(g.season) === latestSeason)

  const windows: [HitRateWindow, number | null][] = [
    ['last_3', 3], ['last_5', 5], ['last_10', 10], ['season', null], ['career', null],
  ]

  const out: HitRate[] = []
  for (const [key, size] of windows) {
    let sample: number[]
    if (key === 'career') sample = values
    else if (key === 'season') sample = values.filter((_, i) => seasonMask[i])
    else sample = values.slice(-(size as number))

    if (sample.length === 0) continue
    const hits = sample.filter((v) => v >= line).length
    out.push({
      window: key,
      games: sample.length,
      hits,
      rate: hits / sample.length,
      average: sample.reduce((a, b) => a + b, 0) / sample.length,
      // Which season "season" is, so the panel can say so.
      ...(key === 'season' ? { season: latestSeason } : {}),
    })
  }
  return out
}

// ---------------------------------------------------------------------------
// Trained model inference
// ---------------------------------------------------------------------------

/**
 * The offseason-aware EWMA the trained model learned from (_prior_ewma), seen
 * from the next game: a game's age is games back plus OFFSEASON_GAP_GAMES for
 * every offseason in between. With no gap this is pandas' ewm(halflife).
 */
function ewmaLast(values: number[], seasons: number[], halfLife: number): number {
  if (values.length === 0) return 0
  const last = seasons[seasons.length - 1]
  let numerator = 0
  let denominator = 0
  for (let i = 0; i < values.length; i++) {
    const age = values.length - 1 - i + OFFSEASON_GAP_GAMES * (last - seasons[i])
    const weight = Math.pow(0.5, age / halfLife)
    numerator += weight * values[i]
    denominator += weight
  }
  return numerator / denominator
}

function mlFeatures(
  games: GameRow[], opponent: string, stat: string, model: TrainedModelFile, aggregates: Aggregates,
): number[] | null {
  const values = games.map((g) => {
    const v = g[stat]
    return typeof v === 'number' && !Number.isNaN(v) ? v : 0
  })
  if (values.length === 0) return null

  const last = games[games.length - 1]
  const seasons = games.map((g) => Number(g.season))
  const position = String(last.position ?? '')
  // The next game's week: a player last seen in an earlier season opens this
  // one in week 1, not week 19. Mirrors features_for_next_game().
  const nextWeek = Number(last.season) < aggregates.constants.current_season ? 1 : Number(last.week) + 1

  const row: Record<string, number> = {
    form_short: ewmaLast(values, seasons, 3),
    form_long: ewmaLast(values, seasons, 8),
    career_avg: values.reduce((a, b) => a + b, 0) / values.length,
    career_std: values.length > 1 ? sampleStd(values) : 0,
    games_played: values.length,
    week: nextWeek,
  }

  // What this defense allowed to the player's position over its last
  // rolling_games games, across seasons — the number the matchup weight reads.
  const allowed = aggregates.position_allowed[`${opponent}|${position}|${stat}`]
  row.def_allowed = typeof allowed === 'number' ? allowed : row.career_avg

  for (const column of aggregates.constants.usage_columns[stat] ?? []) {
    const usage = games.map((g) => {
      const v = g[column]
      return typeof v === 'number' && !Number.isNaN(v) ? v : 0
    })
    row[`usage_${column}_short`] = ewmaLast(usage, seasons, 3)
    row[`usage_${column}_long`] = ewmaLast(usage, seasons, 8)
  }

  const vector: number[] = []
  for (const feature of model.features) {
    if (!(feature in row)) return null
    vector.push(row[feature])
  }
  return vector
}

function mlPredict(model: TrainedModelFile, features: number[]): number {
  let sum = model.weights[0] // intercept
  for (let i = 0; i < features.length; i++) {
    sum += model.weights[i + 1] * ((features[i] - model.mean[i]) / model.scale[i])
  }
  return Math.max(sum, 0)
}

function residualsFor(model: TrainedModelFile, prediction: number): number[] {
  // np.digitize(prediction, bin_edges) — number of edges strictly below.
  let index = 0
  for (const edge of model.bin_edges) if (prediction >= edge) index++
  return model.residual_percentiles[clamp(index, 0, model.residual_percentiles.length - 1)]
}

function mlProbOver(model: TrainedModelFile, prediction: number, line: number): number {
  // Residuals ship as sorted percentiles, so the share of outcomes at or above
  // the line is found by locating (line - prediction) in that ladder.
  const ladder = residualsFor(model, prediction)
  const target = line - prediction
  let below = 0
  for (const r of ladder) if (r < target) below++
  return clamp(1 - below / ladder.length, 0, 1)
}

function mlSpread(model: TrainedModelFile, prediction: number): number {
  const ladder = residualsFor(model, prediction)
  const mean = ladder.reduce((a, b) => a + b, 0) / ladder.length
  return Math.sqrt(ladder.reduce((acc, r) => acc + (r - mean) ** 2, 0) / ladder.length)
}

// ---------------------------------------------------------------------------

export interface ProjectInput {
  player: string
  opponent: string
  stat: string
  line: number
  model: string
  games: GameRow[]
  aggregates: Aggregates
  models: ModelsFile
}

/** Full projection: the chosen model's read plus every alternative. */
export function project(input: ProjectInput): ProjectionResponse {
  const { player, opponent, stat, line, games, aggregates, models } = input
  const { constants } = aggregates

  if (!constants.stat_map[stat]) throw new Error(`Unsupported stat category '${stat}'`)

  // Every game the player has, weighted by age rather than cut off at a window:
  // weights halve every HALF_LIFE_GAMES games back, and each offseason in
  // between adds OFFSEASON_GAP_GAMES. Recent form leads, but a proven career
  // never drops to zero. Mirrors _window() in core/projection.py.
  const ordered = chronological(games)
  const played = ordered.filter((g) => typeof g[stat] === 'number' && !Number.isNaN(g[stat]))
  const values = played.map((g) => g[stat] as number)
  if (values.length === 0) {
    throw new Error(`Not enough recent games for ${player} to run a projection`)
  }

  const weights = decayWeights(played.map((g) => Number(g.season)))
  const shift = createWeight(ordered, opponent, stat, aggregates)
  const { mean: rawMean, ess } = weightedMoments(values, weights)

  // Every probability goes through the shipped calibration map first.
  const alternatives: Record<string, number> = {}
  for (const key of MODEL_KEYS) {
    alternatives[key] = calibrated(models, stat, key, runModel(key, values, weights, line, stat, shift).probOver)
  }

  // The trained model, when it applies to this stat and player.
  const trained = models.models[stat]
  let ml: { projection: number; probOver: number; std: number } | null = null
  if (trained) {
    const features = mlFeatures(ordered, opponent, stat, trained, aggregates)
    if (features) {
      const prediction = mlPredict(trained, features)
      ml = {
        projection: prediction,
        probOver: calibrated(models, stat, 'ml', mlProbOver(trained, prediction, line)),
        std: mlSpread(trained, prediction),
      }
      alternatives.ml = ml.probOver
    }
  }

  let projection: number
  let probOver: number
  let std: number
  let modelKey: string
  let modelLabel: string

  if (input.model === 'ml') {
    if (!ml) {
      throw new Error(
        `The trained model isn't available for ${stat} — not enough history for this player or stat.`,
      )
    }
    projection = ml.projection
    probOver = ml.probOver
    std = ml.std
    modelKey = 'ml'
    modelLabel = 'Trained ridge regression'
  } else {
    const result = runModel(input.model, values, weights, line, stat, shift)
    const calKey = input.model === 'poisson'
      ? 'negbin'
      : (MODEL_KEYS as readonly string[]).includes(input.model) ? input.model : 'ensemble'
    projection = result.projection
    probOver = calibrated(models, stat, calKey, result.probOver)
    std = result.std
    modelKey = result.model
    modelLabel = result.label
  }

  return {
    player, opponent, stat, line,
    projection,
    prob_over: probOver,
    prob_under: 1 - probOver,
    weight: shift,
    model: modelKey,
    model_label: modelLabel,
    form_average: rawMean,
    season_average: values.reduce((a, b) => a + b, 0) / values.length,
    recent_games: values.length,
    effective_games: ess,
    std_dev: std,
    window_games: values.length,
    alternatives,
    hit_rates: hitRates(ordered, stat, line),
    ml_projection: ml ? ml.projection : null,
  }
}

export { quantileFromSorted }
